File: vibebot/plugins/news/og_image.py
import re
from typing import Optional

import requests


def fetch_og_image(url: str, timeout: int = 5) -> Optional[str]:
    """Fetch the Open Graph image URL from an article page. Returns None on failure."""
    try:
        resp = requests.get(
            url,
            timeout=timeout,
            headers={"User-Agent": "VibeBot/1.0"},
            allow_redirects=True,
        )
        if resp.status_code != 200:
            return None
        for pattern in [
            r'<meta[^>]+property=["\']og:image["\'][^>]+content=["\']([^"\']+)["\']',
            r'<meta[^>]+content=["\']([^"\']+)["\'][^>]+property=["\']og:image["\']',
            r'<meta[^>]+name=["\']twitter:image["\'][^>]+content=["\']([^"\']+)["\']',
            r'<meta[^>]+content=["\']([^"\']+)["\'][^>]+name=["\']twitter:image["\']',
        ]:
            match = re.search(pattern, resp.text, re.IGNORECASE)
            if match:
                return match.group(1)
        return None
    except Exception:
        return None

File: src/plugins/news/ogImage.ts
export async function fetchOgImage(url: string, timeoutMs: number = 5000): Promise<string | null> {
  try {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);

    const response = await fetch(url, {
      signal: controller.signal,
      headers: { 'User-Agent': 'VibeBot/1.0' },
      redirect: 'follow',
    });
    clearTimeout(timer);

    if (!response.ok) return null;

    const html = await response.text();
    const patterns = [
      /<meta[^>]+property=["']og:image["'][^>]+content=["']([^"']+)["']/i,
      /<meta[^>]+content=["']([^"']+)["'][^>]+property=["']og:image["']/i,
      /<meta[^>]+name=["']twitter:image["'][^>]+content=["']([^"']+)["']/i,
      /<meta[^>]+content=["']([^"']+)["'][^>]+name=["']twitter:image["']/i,
    ];

    for (const pattern of patterns) {
      const match = html.match(pattern);
      if (match) return match[1];
    }
    return null;
  } catch {
    return null;
  }
}
